pkg/controller/util/endpoint/rate_limiting_queue.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package endpoint
+
+import (
+	"sync"
+
+	"k8s.io/client-go/util/workqueue"
+)
+
+// RateLimitingQueue implements a wrapper on workqueue.RateLimitingInterface that also rate limits initial key additions
+// to the queue (unlike workqueue.RateLimitingInterface which rate limits only error retries).
+// TODO(mborsz): Rename to something that doesn't collide with workqueue.RateLimitingInterface (or refactor workqueue.RateLimitingInterface).
+// TODO(mborsz): Promote to workqueue package once we validate API in endpoint controllers.
+type RateLimitingQueue struct {
+	queue       workqueue.RateLimitingInterface
+	rateLimiter workqueue.RateLimiter
+
+	queuedNotStarted     map[interface{}]bool
+	queuedNotStartedLock sync.Mutex
+}
+
+// NewRateLimitingQueue creates a new RateLimitingQueue instance.
+func NewRateLimitingQueue(queue workqueue.RateLimitingInterface, rateLimiter workqueue.RateLimiter) *RateLimitingQueue {
+	return &RateLimitingQueue{
+		queue:            queue,
+		rateLimiter:      rateLimiter,
+		queuedNotStarted: make(map[interface{}]bool),
+	}
+}
+
+// Enqueue adds key to the queue if it's not there yet.
+func (r *RateLimitingQueue) Enqueue(key interface{}, allowRateLimiting bool) {
+	r.queuedNotStartedLock.Lock()
+	defer r.queuedNotStartedLock.Unlock()
+	if _, ok := r.queuedNotStarted[key]; ok {
+		// The key is waiting in queue.
+		// There is no point in readding the key, so let's save some rate limiter tokens.
+		return
+	}
+	r.queuedNotStarted[key] = true
+	if allowRateLimiting {
+		// We consume rate limiter's token only if we really add element to the queue.
+		r.queue.AddAfter(key, r.rateLimiter.When(key))
+	} else {
+		r.queue.Add(key)
+	}
+}
+
+// EnqueueRetry adds key back to the queue after error occured while handling.
+// Please note that adding back is not subject of rate limiting at RateLimitingQueue level -- it will
+// be rate limited by the queue itself (potentially using different rate limiter).
+func (r *RateLimitingQueue) EnqueueRetry(key interface{}) {
+	r.queue.AddRateLimited(key)
+}
+
+// Get blocks until the next element is available or queue is shutdown.
+func (r *RateLimitingQueue) Get() (interface{}, bool) {
+	key, quit := r.queue.Get()
+
+	r.queuedNotStartedLock.Lock()
```

Comment:
I guess you need to lock the whole Get func.

There may be a race that: 
1. in `Get`, `key, quit := r.queue.Get()` and a key is dequeued
2. While `Enqueue(key)` is also happening and `queuedNotStarted[key]` has key

Reply:
r.queue.Get is a blocking call that can potentially take a long time so taking lock before is not feasible.

But please note that this will work fine in this case: The Enqueue will detect that key hasn't been processed yet (which is correct) and will skip adding an element to queue.

In general, if there are two running Enqueue and Get calls for the same key, the result will depend on the order of taking locks:
* if Enqueue takes lock first, then it will assume that key hasn't been processed yet and skip adding key again to queue, then Get will remove the key from array and pass the key to worker which is correct
* If Get takes lock first, it will remove the key so Enqueue will add element again which is the only safe thing to do as it doesn't know whether Get call has finished and we are in the middle of processing given key or we are still in Get. 

I will try to make it more clear (add some comments and/or rename variable)

The semantics of "key exists in queuedNotStarted" is "given key is either in queue or not passed to the worker yet".
Maybe I can rename this array to something like "notStarted" or "queuedOrNotStarted" to make it more clear.

Reply:
Yeah - it requires some comment.

Reply:
I added comment + renamed